Lib/test/test_capi/test_unicode.py

```diff
@@ -1784,8 +1784,24 @@ def test_widechar(self):
         writer.write_widechar("latin1=\xE9")
         writer.write_widechar("-")
         writer.write_widechar("euro=\u20AC")
+        writer.write_char("-")
+        writer.write_ucs4("max=\U0010ffff", -1)
         writer.write_char('.')
-        self.assertEqual(writer.finish(), "latin1=\xE9-euro=\u20AC.")
+        self.assertEqual(writer.finish(),
+                         "latin1=\xE9-euro=\u20AC-max=\U0010ffff.")
+
+    def test_ucs4(self):
+        writer = self.create_writer(0)
+        writer.write_ucs4("ascii", -1)
```

Comment:
All tests use negative size. I suppose they all will fail.

Please add also tests with embedded null characters, with lone a surrogate pair, with  surrogates, with non null-terminated string (`size` lesser that the string size) and with invalid size.